cpp/ql/src/semmle/code/cpp/ir/internal/CppType.qll

```diff
@@ -159,8 +165,13 @@ private IRType getIRTypeForPRValue(Type type) {
     isUnsignedIntegerType(unspecifiedType) and
     result.(IRUnsignedIntegerType).getByteSize() = type.getSize()
     or
-    unspecifiedType instanceof FloatingPointType and
-    result.(IRFloatingPointType).getByteSize() = type.getSize()
+    exists(FloatingPointType floatType, IRFloatingPointType irFloatType |
+      floatType = unspecifiedType and
+      irFloatType = result and
+      irFloatType.getByteSize() = floatType.getSize() and
+      irFloatType.getBase() = floatType.getBase() and
+      irFloatType.getDomain() = floatType.getDomain()
```

Comment:
Does the QL compiler generate the best possible code for these equalities (a three-column join)? It may not make a big difference here, but it could make a difference for future callers of these getters.

Reply:
It does not, but if I remove the `pragma[noinline]` that you added on `IRFloatingPointType::getByteSize()`, it generates the three-column join you expect. Note that removing that `pragma[noinline]` doesn't seem to produce the bad join order you were seeing when you originally added it, although maybe that bad order was specific to function pointer types.

With the existing `pragma[noinline]`:
```
    {3} r18 = SCAN IRType::IRFloatingPointType#class#ffff AS I OUTPUT I.<0>, I.<2>, I.<3>
    {4} r19 = JOIN r18 WITH IRType::IRFloatingPointType::getByteSize#ff AS R ON FIRST 1 OUTPUT r18.<1>, r18.<0>, r18.<2>, R.<1>
    {4} r20 = JOIN r19 WITH project#Type::FloatingPointType#class#fffff#5_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r19.<2>, r19.<1>, r19.<3>
    {3} r21 = JOIN r20 WITH project#Type::FloatingPointType#class#fffff#4 AS R ON FIRST 2 OUTPUT r20.<0>, r20.<3>, r20.<2>
    {2} r22 = JOIN r21 WITH Type::Type::getSize_dispred#ff AS R ON FIRST 2 OUTPUT r21.<0>, r21.<2>
    {2} r23 = JOIN r22 WITH Type::Type::getUnspecifiedType_dispred#ff_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r22.<1>
    {2} r24 = r17 \/ r23
```

Without the `pragma[noinline]`:
```
    {4} r18 = JOIN dom#IRType::TIRFloatingPointType#fff#shared AS L WITH Type::Type::getSize_dispred#ff AS R ON FIRST 1 OUTPUT R.<1>, L.<0>, L.<1>, L.<2>
    {4} r19 = JOIN r18 WITH IRType::IRNumericType#ff_10#join_rhs AS R ON FIRST 1 OUTPUT r18.<3>, r18.<2>, R.<1>, r18.<1>
    {2} r20 = JOIN r19 WITH IRType::TIRFloatingPointType#ffff_123#join_rhs AS R ON FIRST 3 OUTPUT r19.<3>, r19.<2>
    {2} r21 = JOIN r20 WITH Type::Type::getUnspecifiedType_dispred#ff_10#join_rhs AS R ON FIRST 1 OUTPUT R.<1>, r20.<1>
    {2} r22 = r17 \/ r21
```

Reply:
I see. It doesn't seem very principled to have `noinline` on function pointers but not on floats. I think we have to pick on or the other, and I think the biggest risk is in removing the `noinline`. Is performance and tuple numbers okay in practice if we keep `noinline`?

Reply:
Agreed.
The current join order (with `noinline`) looks reasonable for FP types.